DataFormats/PatCandidates/src/PackedCandidate.cc

```diff
@@ -62,9 +62,12 @@ void pat::PackedCandidate::packVtx(bool unpackAfterwards) {
 }
 
 void pat::PackedCandidate::unpack() const {
-    p4_ = PolarLorentzVector(MiniFloatConverter::float16to32(packedPt_),
+    float pt = MiniFloatConverter::float16to32(packedPt_);
+    double shift = (pt<1. ? 0.1*pt : 0.1/pt); // shift particle phi to break degeneracies in angular separations
+    double phi = int16_t(packedPhi_)*3.2f/std::numeric_limits<int16_t>::max() + shift*3.2/std::numeric_limits<int16_t>::max();
```

Comment:
Where is the degeneracy, actually?
Is it at the distance corresponding to the jet distance parameter (do we
have the largest as 1.2?),
or is it across the board?

If we use 0.1*pt, will it still work for the heavy flavor ID using
ghosts with very low momentum?


On 6/11/15 10:18 PM, Dinko Ferencek wrote:
> In DataFormats/PatCandidates/src/PackedCandidate.cc
> <https://github.com/cms-sw/cmssw/pull/9554#discussion_r32288004>:
> 
>> @@ -62,9 +62,12 @@ void pat::PackedCandidate::packVtx(bool unpackAfterwards) {
>>  }
>>  
>>  void pat::PackedCandidate::unpack() const {
>> -    p4_ = PolarLorentzVector(MiniFloatConverter::float16to32(packedPt_),
>> +    float pt = MiniFloatConverter::float16to32(packedPt_);
>> +    double shift = (pt<1. ? 0.1*pt : 0.1/pt); // shift particle phi to break degeneracies in angular separations
>> +    double phi = int16_t(packedPhi_)*3.2f/std::numeric_limits<int16_t>::max() + shift*3.2/std::numeric_limits<int16_t>::max();
> 
> @slava <https://github.com/slava>, even if we were packing in theta, the
> grid would still exist in eta. Granted, it would not be uniform in eta,
> but degeneracies would still be possible. Furthermore, even if at the
> same time we had a non-linear packing of phi, there would be some
> symmetry among the 4 quadrants in the eta-phi plane and degeneracies
> would in principle still be possible (although I would assume they would
> be significantly less likely than now).
> 
> I don't think it makes sense to pack in cos(phi) since the measured phi
> resolution should be more or less invariant under phi. Hence, packing in
> phi is what makes most sense. Concerning packing in theta, yes, it
> probably makes more sense from the experimental point of view. However,
> constructors of 4-vector classes generally do not take theta as input
> which means that packing and unpacking would involve computationally
> more intensive non-linear functions. So here again I think it makes more
> sense to pack eta directly. In fact, the grid in eta becomes in some
> sense too dense in the forward region so there is no real issue.
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/9554/files#r32288004>.
> 

-- 
***********************************************************************
 Vyacheslav (Slava) Krutelyov
 TAMU: Physics Dept Texas A&M MS4242, College Station, TX 77843-4242
 CERN: 42-R-027
 AIM/Skype: siava16              googleTalk: slava77@gmail.com
 (630) 291-5128  Cell (US)       +41 76 275 7116 Cell (CERN)

-----------------------------------------------------------------------